sources/react/template-all-icons.js

```diff
@@ -0,0 +1,55 @@
+const toString = array => array.reduce( ( acc, item ) => ( acc = acc + "'" + item + "', " ), '' );
+
+const prepareAllIcons = ( {
+	components,
+	iconsThatNeedOffset,
+	iconsThatNeedOffsetX,
+	iconsThatNeedOffsetY,
+} ) => `
+/** @ssr-ready **/
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+export default function( { icon: iconProp, className, onClick, size = 24, ...otherProps } ) {
+
+	const iconsThatNeedOffset = [ ${ toString( iconsThatNeedOffset ) } ];
+
+	const iconsThatNeedOffsetX = [ ${ toString( iconsThatNeedOffsetX ) } ];
+
+	const iconsThatNeedOffsetY = [ ${ toString( iconsThatNeedOffsetY ) } ];
+
+	const doesItNeedOffset = ( name, icons ) => icons.indexOf( name ) >= 0;
+	const isModulo18 = size => size % 18 === 0;
+
+	const icon = 'gridicons-' + iconProp;
+	const needsOffset = doesItNeedOffset( icon, iconsThatNeedOffset ) && isModulo18( size );
+	const needsOffsetX = doesItNeedOffset( icon, iconsThatNeedOffsetX ) && isModulo18( size );
+	const needsOffsetY = doesItNeedOffset( icon, iconsThatNeedOffsetY ) && isModulo18( size );
+
+	let svg;
+	const iconClass = [
+		'gridicon',
+		icon,
+		className,
+		needsOffset ? 'needs-offset' : false,
+		needsOffsetX ? 'needs-offset-x' : false,
+		needsOffsetY ? 'needs-offset-y' : false,
+	].filter( Boolean ).join( ' ' );
+
+	switch ( icon ) {
+		default:
+			svg = <svg height={ size } width={ size } { ...otherProps } />;
+			break;
+
+		${ components }
+
+	}
+
+	return ( svg );
+}
+`;
+
+module.exports = prepareAllIcons;
```

Comment:
What about changing the this template to `import` all individual icon files, and then something like

```jsx
import React from 'react';

${ imports }

export default function( { icon: iconProp, className, onClick, size = 24, ...otherProps } ) {
	const IndividualGridicon = toCamelCase( icon ); 
	return <IndividualGridicon className={ className } onClick={ onClick } size={ size } ...otherProps />;
}
```

(You can then also omit `icon` from the destructuring in the individual templace.)

Reply:
I haven't done it because my understanding is that the CommonJS `index.js` will be bigger: you are embedding react components that contain the svg icon. This way, we directly insert the svg icon.

Reply:
It'd just be nice to get `import { GridiconFoo } from 'gridicons';`. But yeah, let's leave it as-is for now; we can always iterate.